src/test/java/org/jenkinsci/plugins/workflow/cps/DeserializationTest.java

```diff
@@ -0,0 +1,136 @@
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                story.j.jenkins.getWorkspaceFor(p).child(A_GROOVY).write(A_CONTENTS, null);
+                ScriptApproval.get().approveSignature("method groovy.lang.Binding getVariables");
+                story.j.jenkins.getWorkspaceFor(p).child(B_GROOVY).write(B_CONTENTS, null);
+                // Control case:
+                p.setDefinition(new CpsFlowDefinition(LOAD_PREFIX + LOAD_SUFFIX, true));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.assertLogContains("a ran on value from b", b);
+                story.j.assertLogContains("Pre-semaphore value is " + EXISTING_VAR_VALUE, b);
+                // Test case:
+                p.setDefinition(new CpsFlowDefinition(LOAD_PREFIX + LOAD_SEMAPHORE + LOAD_SUFFIX, true));
+                b = p.scheduleBuild2(0).getStartCondition().get();
+                SemaphoreStep.waitForStart("wait/1", b);
+                story.j.assertLogContains("Pre-semaphore value is " + EXISTING_VAR_VALUE, b);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+                WorkflowRun b = p.getBuildByNumber(1);
+                SemaphoreStep.success("wait/1", null);
+                story.j.assertLogContains("a ran on value from b", story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b)));
+                story.j.assertLogContains("Post-semaphore value is " + EXISTING_VAR_VALUE, b);
+            }
+        });
+    }
+
+    @Test
+    public void clashingVariablesOnRestart() throws Exception {
+        String clashingValue = "0";
+        String sameVariable = EXISTING_VAR_NAME + " = " + clashingValue + ";";
+        // Testing with identical variable names injected via load step
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                story.j.jenkins.getWorkspaceFor(p).child(A_GROOVY).write(A_CONTENTS, null);
+                ScriptApproval.get().approveSignature("method groovy.lang.Binding getVariables");
+                story.j.jenkins.getWorkspaceFor(p).child(B_GROOVY).write(B_CONTENTS, null);
+                // Control case:
+                p.setDefinition(new CpsFlowDefinition(sameVariable + LOAD_PREFIX + LOAD_SUFFIX, true));
+                WorkflowRun b = story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+                story.j.assertLogContains("a ran on value from b", b);
+                story.j.assertLogContains("Pre-semaphore value is " + clashingValue, b);
```

Comment:
Just for consistency:

```suggestion
                story.j.assertLogContains("Pre-semaphore value is " + clashingValue, b);
                story.j.assertLogContains("Post-semaphore value is " + clashingValue, b);
```